filebeat/reader/readfile/decoder.go

```diff
@@ -0,0 +1,147 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package readfile
+
+import (
+	"io"
+
+	"golang.org/x/text/encoding"
+	"golang.org/x/text/transform"
+
+	"github.com/elastic/beats/libbeat/common/streambuf"
+)
+
+// decoderReader reads from an input file and converts its content into UTF-8.
+// It stores the size of the original bytes, so offset can be calculated later in
+// the line scanner.
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	offset     int // segment byte offset
+
+	bufferSize int
+}
+
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) (*decoderReader, error) {
+	r := &decoderReader{
+		in:         in,
+		decoder:    codec.NewDecoder(),
+		buf:        streambuf.New(nil),
+		encodedBuf: streambuf.New(nil),
+		offset:     0, // TODO init correctly
+		bufferSize: bufferSize,
+	}
+
+	return r, nil
+}
+
+// seekToLastRead reads from the input file until the saved offset is reached.
+// It is the substitution of seeking in the file.
+// This function is no longer required after registry refactoring is done.
+func (r *decoderReader) seekToLastRead() error {
+	if r.offset == 0 {
+		return nil
+	}
+
+	remaining := r.offset
+	for remaining > 0 {
+		s := r.bufferSize
+		if remaining < s {
+			s = remaining
+		}
+
+		b := make([]byte, s)
```

Comment:
In case this allocates, we should move the allocation before the loop (check escape analysis, [See blog by rakyll](https://rakyll.org/codegen/)) and use `r.in.Read(b[:s])`.